JMMServer/Tasks/RecreateAllGroupsTask.cs

```diff
@@ -0,0 +1,373 @@
+            }
+
+            using (ITransaction trans = session.BeginTransaction())
+            {
+                _animeGroupRepo.InsertBatch(session, newGroupsToSeries.Select(gts => gts.Item1));
+                trans.Commit();
+            }
+
+            // Anime groups should have IDs now they've been inserted. Now assign the group ID's to their respective series
+            // (The caller of this method will be responsible for saving the AnimeSeries)
+            foreach (Tuple<AnimeGroup, AnimeSeries> groupAndSeries in newGroupsToSeries)
+            {
+                groupAndSeries.Item2.AnimeGroupID = groupAndSeries.Item1.AnimeGroupID;
+            }
+
+            _log.Info("Generated {0} AnimeGroups", newGroupsToSeries.Length);
+
+            return newGroupsToSeries.Select(gts => gts.Item1);
+        }
+
+        /// <summary>
+        /// Creates <see cref="AnimeGroup"/> that contain <see cref="AnimeSeries"/> that appear to be related.
+        /// </summary>
+        /// <remarks>
+        /// This method assumes that there are no active transactions on the specified <paramref name="session"/>.
+        /// </remarks>
+        /// <param name="session">The NHibernate session.</param>
+        /// <param name="seriesList">The list of <see cref="AnimeSeries"/> to create groups for.</param>
+        /// <returns>A sequence of the created <see cref="AnimeGroup"/>s.</returns>
+        private IEnumerable<AnimeGroup> AutoCreateGroupsWithRelatedSeries(ISessionWrapper session, IReadOnlyCollection<AnimeSeries> seriesList)
+        {
+            _log.Info("Auto-generating AnimeGroups for {0} AnimeSeries based on aniDB relationships", seriesList.Count);
+
+            DateTime now = DateTime.Now;
+            Regex truncateYearRegex = new Regex(@"\s*\(\d{4}\)$");
+            var grpCalculator = AutoAnimeGroupCalculator.Create(session);
+
+            _log.Info("The following exclusions will be applied when generating the groups: " + grpCalculator.Exclusions);
+
+            // Group all of the specified series into their respective groups (keyed by the groups main anime ID)
+            var seriesByGroup = seriesList.ToLookup(s => grpCalculator.GetGroupAnimeId(s.AniDB_ID));
+            var newGroupsToSeries = new List<Tuple<AnimeGroup, IReadOnlyCollection<AnimeSeries>>>(seriesList.Count);
+
+            foreach (var groupAndSeries in seriesByGroup)
+            {
+                int mainAnimeId = groupAndSeries.Key;
+                AnimeSeries mainSeries = groupAndSeries.FirstOrDefault(series => series.AniDB_ID == mainAnimeId);
+                AnimeGroup animeGroup = new AnimeGroup();
+                string groupName = null;
+
+                if (mainSeries != null)
+                {
+                    animeGroup.Populate(mainSeries, now);
+                    groupName = mainSeries.GetSeriesName(session);
+                }
+                else // The anime chosen as the group's main anime doesn't actually have a series

```

Comment:
Ah, there is actually a reason for that. I should have explained how my auto grouping works. Before I get to that, I'll mention why I modified the algorithm in the first place.
When I ran the existing implementation on my database, I ended up with groups such as:
- Initial D Fifth Stage
- Yuru Yuri 2
- Minami-ke Omatase
- Chihayafuru 2
- Durarara!!x2 Shou
- Gekijouban Toaru Majutsu no Index: Endymion no Kiseki
etc.

Which didn't seem right to me. My version basically builds the groups as needed and tries to pick a "master" anime to use for generating each group name (NOTE: It doesn't look at the series, but looks at the AniDB_Anime and relations). For the same database I ended up with the below anime chosen for the group names (with the series I actually have in JMM. I only have series from the last few years in my JMM database, everything else was in aniDB already and sitting on a file server):
- Initial D
  - Initial D Fifth Stage
  - Initial D Final Stage
- Yuru Yuri
  - Yuru Yuri 2
  - Yuru Yuri Nachuyachumi!
  - Yuru Yuri San Hai!
  - Yuru Yuri Nachuyachumi!+
- Minami-ke
  - Minami-ke Tadaima
  - Minami-ke Omatase
- Chihayafuru
  - Chihayafuru 2
  - Chihayafuru 2: Waga Mi Yo ni Furu Nagame Seshi Ma ni
- Durarara!!
  - Durarara!!x2 Shou
  - Durarara!!x2 Ten
  - Durarara!!x2 Ketsu
- Toaru Majutsu no Index
  - Toaru Majutsu no Index II
  - Toaru Kagaku no Railgun OVA
  - Gekijouban Toaru Majutsu no Index: Endymion no Kiseki
  - Toaru Kagaku no Railgun S
  - Toaru Kagaku no Railgun S: Daiji na Koto wa Zenbu Sentou ni Osowatta

So as you can see, for my database, it has actually chosen to use Anime that I don't actually have a series of (hence the reason for the fallback). I can try and explain how it works further if you want me to.
I've uploaded a simple test project that I used for experimentation purposes here:
[pmcleish/JMMAutoGroupTest](https://github.com/pmcleish/JMMAutoGroupTest)
You might be able to play around with that to see how it works as well.